aws-android-sdk-s3/src/test/java/com/amazonaws/mobileconnectors/s3/transferutility/UploadInputStreamTest.java

```diff
@@ -0,0 +1,221 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazonaws.mobileconnectors.s3.transferutility;
+
+import android.content.Context;
+import android.util.Log;
+
+import androidx.test.platform.app.InstrumentationRegistry;
+
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.CannedAccessControlList;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.UUID;
+
+import static org.junit.Assert.assertFalse;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Matchers.isNull;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@RunWith(RobolectricTestRunner.class)
+public class UploadInputStreamTest {
+
+    private static final String TAG = "InputStreamTest";
+
+    private static final String DEFAULT_BUCKET = "s3-upload-input-stream-unit-test-" + getRandomString();
+    private final InputStream inputStream;
+    private final ObjectMetadata metadata;
+    private final CannedAccessControlList cannedAcl;
+    private final TransferListener listener;
+    private final TransferUtility transferUtility;
+
+    public UploadInputStreamTest() {
+        AmazonS3Client s3 = mock(AmazonS3Client.class);
+        Context context = InstrumentationRegistry.getInstrumentation().getContext();
+        transferUtility = spy(TransferUtility.builder()
+                .defaultBucket(DEFAULT_BUCKET)
+                .s3Client(s3)
+                .context(context)
+                .build());
+        inputStream = new ByteArrayInputStream("test data".getBytes());
+        metadata = new ObjectMetadata();
+        cannedAcl = CannedAccessControlList.Private;
+        listener = new UploadListener();
+    }
+
+    // Testing the upload(String key, InputStream inputStream, UploadOptions options) method calls
+    // upload(String bucket, String key, File file, ObjectMetadata metadata,
+    // CannedAccessControlList cannedAcl, TransferListener listener) in fact.
+    @Test
+    public void testUpload() throws IOException {
+        String key = getRandomString();
+        transferUtility.upload(key, inputStream, UploadOptions.builder().build());
+        verify(transferUtility, times(1)).upload(eq(DEFAULT_BUCKET), eq(key), any(File.class),
+                any(ObjectMetadata.class), isNull(CannedAccessControlList.class), isNull(TransferListener.class));
+    }
+
+    @Test
+    public void testUploadWithBucket() throws IOException {
```

Comment:
I rarely say there are too many tests, but in this case, I think there might be :).  Lots of testing is good, but we also want to keep in mind the maintainability of our test code.   For example, if we added a 5th parameter to the `Builder` in the future, we'd have to add 5 more tests if we want to follow the pattern that you've started here (and a 6th parameter would require 6 more tests, etc), but we really don't need that many to validate that our code works as expected.  As written, there is a lot of duplicate coverage of the same code paths.    Instead of testing every possible permutation, I think you could have three tests:

1. Test that `upload(String key, String inputStream)` calls the `File` based upload method with all of the expected defaults.  (It doesn't look like you have a test for this method at all right now actually)
2. Test that `upload(String key, String inputStream, UploadOptions options)` with a default Builder calls the `File` based upload method with all of the expected defaults.
3. Test that `upload(String key, String inputStream, UploadOptions options)` with all parameters specified calls the `File` based upload method with the same parameters that were provided as input.